File: backend/infrastructure/repositories/interview_repository.py
import json
from uuid import UUID

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import selectinload

from domain.answer import Answer, AnswerStatus
from domain.interview import Interview
from domain.question import Question, QuestionType
from domain.topic import TopicEntry, TopicStatus
from infrastructure.db.models import AnswerModel, InterviewModel, QuestionModel
from infrastructure.db.session import get_session_factory
from infrastructure.repositories.base import InterviewRepositoryInterface
from infrastructure.repositories.topic_serialization import (
    serialize_topic_plan as _serialize_topic_plan,
    deserialize_topic_plan as _deserialize_topic_plan,
)


class PostgresInterviewRepository(InterviewRepositoryInterface):
    def __init__(self):
        self._get_session_factory = get_session_factory

    async def save(self, interview: Interview) -> None:
        async with self._get_session_factory()() as session:
            async with session.begin():
                existing = await session.get(InterviewModel, interview.id)

                if existing:
                    existing.user_id = interview.user_id
                    existing.candidate_name = interview.candidate_name
                    existing.job_role = interview.job_role
                    existing.status = interview.status.value
                    existing.resume_id = interview.resume_id
                    existing.resume_snapshot = interview.resume_snapshot
                    existing.jd_snapshot = interview.jd_snapshot
                    existing.current_question_index = interview.current_question_index
                    existing.total_questions = interview.total_questions
                    existing.topic_plan = _serialize_topic_plan(interview.topic_plan)
                    existing.current_topic_id = interview.current_topic_id
                    existing.analysis = json.dumps(interview.analysis) if interview.analysis else None

                    # Batch-load existing question IDs and answer question_ids
                    q_ids = [q.id for q in interview.questions]
                    existing_q_result = await session.execute(
                        select(QuestionModel.id).where(
                            QuestionModel.interview_id == interview.id,
                            QuestionModel.id.in_(q_ids),
                        )
                    )
                    existing_q_ids = {row[0] for row in existing_q_result.all()}

                    a_q_ids = [interview.questions[idx].id for idx in interview.answers if idx < len(interview.questions)]
                    existing_a_result = await session.execute(
                        select(AnswerModel.question_id).where(
                            AnswerModel.interview_id == interview.id,
                            AnswerModel.question_id.in_(a_q_ids),
                        )
                    )
                    existing_a_q_ids = {row[0] for row in existing_a_result.all()}

                    # Insert new questions
                    for q in interview.questions:
                        if q.id not in existing_q_ids:
                            session.add(
                                QuestionModel(
                                    id=q.id,
                                    interview_id=interview.id,
                                    question_text=q.text,
                                    question_index=q.order,
                                    question_type=q.question_type.value,
                                )
                            )

                    # Insert/update answers
                    for idx, a in interview.answers.items():
                        if idx < len(interview.questions):
                            q_id = interview.questions[idx].id
                            if q_id in existing_a_q_ids:
                                # Update existing answer
                                a_row_result = await session.execute(
                                    select(AnswerModel).where(
                                        AnswerModel.interview_id == interview.id,
                                        AnswerModel.question_id == q_id,
                                    )
                                )
                                a_row = a_row_result.scalar_one()
                                a_row.transcript = a.transcript
                                a_row.answer_status = a.answer_status.value if a.answer_status else None
                            else:
                                session.add(
                                    AnswerModel(
                                        interview_id=interview.id,
                                        question_id=q_id,
                                        transcript=a.transcript,
                                        answer_status=a.answer_status.value if a.answer_status else None,
                                    )
                                )
                else:
                    db_interview = InterviewModel(
                        id=interview.id,
                        user_id=interview.user_id,
                        candidate_name=interview.candidate_name,
                        job_role=interview.job_role,
                        status=interview.status.value,
                        resume_id=interview.resume_id,
                        resume_snapshot=interview.resume_snapshot,
                        jd_snapshot=interview.jd_snapshot,
                        current_question_index=interview.current_question_index,
                        total_questions=interview.total_questions,
                        topic_plan=_serialize_topic_plan(interview.topic_plan),
                        current_topic_id=interview.current_topic_id,
                        analysis=json.dumps(interview.analysis) if interview.analysis else None,
                    )
                    session.add(db_interview)

                    for q in interview.questions:
                        session.add(
                            QuestionModel(
                                id=q.id,
                                interview_id=interview.id,
                                question_text=q.text,
                                question_index=q.order,
                                question_type=q.question_type.value,
                            )
                        )

                    for idx, a in interview.answers.items():
                        if idx < len(interview.questions):
                            session.add(
                                AnswerModel(
                                    interview_id=interview.id,
                                    question_id=interview.questions[idx].id,
                                    transcript=a.transcript,
                                    answer_status=a.answer_status.value if a.answer_status else None,
                                )
                            )

    async def get(self, interview_id: UUID) -> Interview | None:
        async with self._get_session_factory()() as session:
            result = await session.execute(
                select(InterviewModel)
                .where(InterviewModel.id == interview_id)
                .options(
                    selectinload(InterviewModel.questions),
                    selectinload(InterviewModel.answers),
                )
            )
            db_interview = result.scalar_one_or_none()

            if not db_interview:
                return None

            return self._to_domain(db_interview)

    async def list_all(self) -> list[Interview]:
        async with self._get_session_factory()() as session:
            result = await session.execute(
                select(InterviewModel).options(
                    selectinload(InterviewModel.questions),
                    selectinload(InterviewModel.answers),
                )
            )
            db_interviews = result.scalars().all()
            return [self._to_domain(i) for i in db_interviews]

    async def delete(self, interview_id: UUID) -> bool:
        async with self._get_session_factory()() as session:
            async with session.begin():
                db_interview = await session.get(InterviewModel, interview_id)
                if not db_interview:
                    return False
                await session.delete(db_interview)
                return True

    def _to_domain(self, db_interview: InterviewModel) -> Interview:
        from domain.interview_state import InterviewState

        questions = sorted(db_interview.questions, key=lambda q: q.question_index)
        domain_questions = [
            Question(
                id=q.id,
                text=q.question_text,
                question_type=QuestionType(q.question_type),
                order=q.question_index,
            )
            for q in questions
        ]

        answers_dict: dict[int, Answer] = {}
        q_id_to_index = {q.id: q.question_index for q in questions}
        for a in db_interview.answers:
            if a.question_id in q_id_to_index:
                idx = q_id_to_index[a.question_id]
                answers_dict[idx] = Answer(
                    question_id=a.question_id,
                    transcript=a.transcript,
                    answer_status=AnswerStatus(a.answer_status) if a.answer_status else None,
                )

        topic_plan = _deserialize_topic_plan(db_interview.topic_plan) if db_interview.topic_plan else []

        return Interview(
            id=db_interview.id,
            user_id=db_interview.user_id,
            candidate_name=db_interview.candidate_name,
            job_role=db_interview.job_role,
            status=InterviewState(db_interview.status),
            resume_id=db_interview.resume_id,
            resume_snapshot=db_interview.resume_snapshot or "",
            jd_snapshot=db_interview.jd_snapshot or "",
            questions=domain_questions,
            answers=answers_dict,
            current_question_index=db_interview.current_question_index,
            total_questions=db_interview.total_questions,
            topic_plan=topic_plan,
            current_topic_id=db_interview.current_topic_id,
            analysis=json.loads(db_interview.analysis) if db_interview.analysis else None,
        )

File: backend/src/repositories/interview-repository.js
import { getPool } from "../infrastructure/db/pool.js";
import { withTransaction } from "../infrastructure/db/transaction.js";
import { Interview } from "../models/interview.js";
import { Question } from "../models/question.js";
import { Answer } from "../models/answer.js";
import { AnswerStatus } from "../models/answer-status.js";
import { InterviewState } from "../models/interview-state.js";
import { InterviewMode } from "../models/interview-mode.js";
import { QuestionType } from "../models/question-type.js";
import { AnalysisStatus } from "../models/analysis-status.js";
import { TopicEntry } from "../models/topic.js";
import { TopicStatus } from "../models/topic-status.js";

function serializeTopicPlan(topics) {
  return JSON.stringify(
    topics.map((t) => ({
      id: t.id,
      label: t.label,
      source: t.source,
      primary_question: t.primaryQuestion,
      priority: t.priority,
      status: t.status,
      questions_asked: t.questionsAsked,
      exhaustion_reason: t.exhaustionReason,
    })),
  );
}

function deserializeTopicPlan(raw) {
  if (!raw) return [];
  const arr = JSON.parse(raw);
  return arr.map(
    (t) =>
      new TopicEntry({
        id: t.id,
        label: t.label,
        source: t.source || "",
        primaryQuestion: t.primary_question || "",
        priority: t.priority || 5,
        status: TopicStatus[t.status || "AVAILABLE"],
        questionsAsked: t.questions_asked || 0,
        exhaustionReason: t.exhaustion_reason || null,
      }),
  );
}

export class PostgresInterviewRepository {
  async save(interview) {
    await withTransaction(async (client) => {
      const existing = await client.query("SELECT id FROM interviews WHERE id = $1", [interview.id]);

      if (existing.rows.length > 0) {
        await client.query(
          `UPDATE interviews SET
            user_id = $1, candidate_name = $2, job_role = $3, interview_mode = $4,
            status = $5, resume_id = $6, resume_snapshot = $7, jd_snapshot = $8,
            current_question_index = $9, total_questions = $10, topic_plan = $11,
            current_topic_id = $12, analysis = $13, analysis_status = $14
           WHERE id = $15`,
          [
            interview.userId,
            interview.candidateName,
            interview.jobRole,
            interview.interviewMode,
            interview.status,
            interview.resumeId,
            interview.resumeSnapshot,
            interview.jdSnapshot,
            interview.currentQuestionIndex,
            interview.totalQuestions,
            serializeTopicPlan(interview.topicPlan),
            interview.currentTopicId,
            interview.analysis ? JSON.stringify(interview.analysis) : null,
            interview.analysisStatus,
            interview.id,
          ],
        );

        const qIds = interview.questions.map((q) => q.id);
        const existingQResult = await client.query(
          "SELECT id FROM questions WHERE interview_id = $1 AND id = ANY($2)",
          [interview.id, qIds],
        );
        const existingQIds = new Set(existingQResult.rows.map((r) => r.id));

        for (const q of interview.questions) {
          if (!existingQIds.has(q.id)) {
            await client.query(
              `INSERT INTO questions (id, interview_id, question_text, question_index, question_type, created_at)
               VALUES ($1, $2, $3, $4, $5, NOW())`,
              [q.id, interview.id, q.text, q.order, q.questionType],
            );
          }
        }

        const answerQIds = Object.entries(interview.answers)
          .filter(([idx]) => Number(idx) < interview.questions.length)
          .map(([idx]) => interview.questions[Number(idx)].id);
        const existingAResult = await client.query(
          "SELECT question_id FROM answers WHERE interview_id = $1 AND question_id = ANY($2)",
          [interview.id, answerQIds],
        );
        const existingAnswerQIds = new Set(existingAResult.rows.map((r) => r.question_id));

        for (const [idx, a] of Object.entries(interview.answers)) {
          const index = Number(idx);
          if (index < interview.questions.length) {
            const qId = interview.questions[index].id;
            if (existingAnswerQIds.has(qId)) {
              await client.query(
                "UPDATE answers SET transcript = $1, answer_status = $2 WHERE interview_id = $3 AND question_id = $4",
                [a.transcript, a.answerStatus || null, interview.id, qId],
              );
            } else {
              await client.query(
                `INSERT INTO answers (id, interview_id, question_id, transcript, answer_status, created_at)
                 VALUES (gen_random_uuid(), $1, $2, $3, $4, NOW())`,
                [interview.id, qId, a.transcript, a.answerStatus || null],
              );
            }
          }
        }
      } else {
        await client.query(
          `INSERT INTO interviews (id, user_id, candidate_name, job_role, interview_mode, status,
            resume_id, resume_snapshot, jd_snapshot, current_question_index, total_questions,
            topic_plan, current_topic_id, analysis, analysis_status, created_at)
           VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15, NOW())`,
          [
            interview.id,
            interview.userId,
            interview.candidateName,
            interview.jobRole,
            interview.interviewMode,
            interview.status,
            interview.resumeId,
            interview.resumeSnapshot,
            interview.jdSnapshot,
            interview.currentQuestionIndex,
            interview.totalQuestions,
            serializeTopicPlan(interview.topicPlan),
            interview.currentTopicId,
            interview.analysis ? JSON.stringify(interview.analysis) : null,
            interview.analysisStatus,
          ],
        );

        for (const q of interview.questions) {
          await client.query(
            `INSERT INTO questions (id, interview_id, question_text, question_index, question_type, created_at)
             VALUES ($1, $2, $3, $4, $5, NOW())`,
            [q.id, interview.id, q.text, q.order, q.questionType],
          );
        }

        for (const [idx, a] of Object.entries(interview.answers)) {
          const index = Number(idx);
          if (index < interview.questions.length) {
            await client.query(
              `INSERT INTO answers (id, interview_id, question_id, transcript, answer_status, created_at)
               VALUES (gen_random_uuid(), $1, $2, $3, $4, NOW())`,
              [interview.id, interview.questions[index].id, a.transcript, a.answerStatus || null],
            );
          }
        }
      }
    });
  }

  async get(interviewId) {
    const pool = getPool();
    const result = await pool.query("SELECT * FROM interviews WHERE id = $1", [interviewId]);
    const row = result.rows[0];
    if (!row) return null;

    const [qResult, aResult] = await Promise.all([
      pool.query(
        "SELECT id, question_text, question_index, question_type FROM questions WHERE interview_id = $1 ORDER BY question_index",
        [interviewId],
      ),
      pool.query(
        "SELECT id, question_id, transcript, answer_status FROM answers WHERE interview_id = $1",
        [interviewId],
      ),
    ]);

    const questions = qResult.rows.map(
      (q) =>
        new Question({
          id: q.id,
          text: q.question_text,
          questionType: QuestionType[q.question_type || "PRIMARY"],
          order: q.question_index,
        }),
    );

    const qIdToIndex = new Map();
    for (const q of questions) {
      qIdToIndex.set(q.id, q.order);
    }

    const answers = {};
    for (const a of aResult.rows) {
      const idx = qIdToIndex.get(a.question_id);
      if (idx !== undefined) {
        answers[idx] = new Answer({
          questionId: a.question_id,
          transcript: a.transcript || "",
          answerStatus: a.answer_status
            ? AnswerStatus[a.answer_status]
            : null,
        });
      }
    }

    const topicPlan = deserializeTopicPlan(row.topic_plan || "[]");

    return new Interview({
      id: row.id,
      userId: row.user_id,
      candidateName: row.candidate_name,
      jobRole: row.job_role,
      interviewMode: InterviewMode[row.interview_mode] || InterviewMode.MIXED,
      status: InterviewState[row.status],
      resumeId: row.resume_id,
      resumeSnapshot: row.resume_snapshot || "",
      jdSnapshot: row.jd_snapshot || "",
      questions,
      answers,
      currentQuestionIndex: row.current_question_index,
      totalQuestions: row.total_questions,
      topicPlan,
      currentTopicId: row.current_topic_id,
      analysis: row.analysis ? JSON.parse(row.analysis) : null,
      analysisStatus:
        AnalysisStatus[row.analysis_status] || AnalysisStatus.PENDING,
      createdAt: row.created_at,
    });
  }

  async listAll() {
    const pool = getPool();
    const result = await pool.query("SELECT id FROM interviews");
    const interviews = [];
    for (const row of result.rows) {
      const interview = await this.get(row.id);
      if (interview) interviews.push(interview);
    }
    return interviews;
  }

  async delete(interviewId) {
    const pool = getPool();
    const result = await pool.query("DELETE FROM interviews WHERE id = $1", [interviewId]);
    return (result.rowCount ?? 0) > 0;
  }
}
